test/rules/avoid_type_to_string.dart

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) 2020, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// test w/ `pub run test -N avoid_type_to_string`
+
+
+// SHARED
+
+class A {
+  String toString() {}
+}
+
+class TypeChildWithOverride extends Type {
+  @override
+  String toString() {}
+}
+class TypeGrandChildWithOverride extends TypeChildWithOverride {}
+
+class TypeChildNoOverride extends Type {}
+class TypeGrandChildNoOverride extends TypeChildNoOverride {}
+
+mixin ToStringMixin {
+  String toString() {}
+}
+
+
+// BAD
+
+class Bad {
+  void doBad() {
+    A().runtimeType.toString(); // LINT
+    TypeChildNoOverride().toString(); // LINT
+    TypeGrandChildNoOverride().toString(); // LINT
+  }
+}
+
+class BadWithType extends Type {
+  void doBad() {
+    toString(); // LINT
+    this.toString(); // LINT
+
+    Future.value('hello').whenComplete(toString); // LINT
```

Comment:
I believe that the issue here is that these expressions appear as arguments and not that they are arguments to `whenComplete`. Consider writing a dummy function like
```
void takesFunction(Function f) {}
```
that's used to test that these expressions are caught.

Reply:
Very good point, that feels much better.